devito/dse/manipulation.py

```diff
@@ -23,35 +30,69 @@ def collect_nested(expr, aggressive=False):
-
-            w_numbers = [i for i in rebuilt if any(j.is_Number for j in i.args)]
-            wo_numbers = [i for i in rebuilt if i not in w_numbers]
-
-            w_numbers = collect_const(expr.func(*w_numbers))
-            wo_numbers = expr.func(*wo_numbers)
+            args, candidates = zip(*[run(arg) for arg in expr.args])
+            candidates = ReducerMap.fromdicts(*candidates)
+
+            funcs = candidates.getall('funcs', [])
+            pows = candidates.getall('pows', [])
+            coeffs = candidates.getall('coeffs', [])
+
+            # Functions/Pows are collected first, coefficients afterwards
+            # Note: below we use sets, but SymPy will ensure determinism
+            args = set(args)
+            w_funcs = {i for i in args if any(j in funcs for j in i.args)}
+            args -= w_funcs
+            w_pows = {i for i in args if any(j in pows for j in i.args)}
+            args -= w_pows
+            w_coeffs = {i for i in args if any(j in coeffs for j in i.args)}
+            args -= w_coeffs
+
+            # Collect common funcs
+            w_funcs = collect(expr.func(*w_funcs), funcs, evaluate=False)
+            try:
+                w_funcs = Add(*[Mul(k, collect_const(v)) for k, v in w_funcs.items()])
+            except AttributeError:
+                assert w_funcs == 0
+
+            # Collect common pows
+            w_pows = collect(expr.func(*w_pows), pows, evaluate=False)
+            try:
+                w_pows = Add(*[Mul(k, collect_const(v)) for k, v in w_pows.items()])
+            except AttributeError:
+                assert w_pows == 0
+
+            # Collect common coefficients
+            w_coeffs = collect_const(expr.func(*w_coeffs))
+
+            rebuilt = Add(w_funcs, w_pows, w_coeffs, *args)
+
+            return rebuilt, {}
+        elif expr.is_Mul:
+            args, candidates = zip(*[run(arg) for arg in expr.args])
 
-            if aggressive is True and wo_numbers:
-                for i in flatten(candidates):
-                    wo_numbers = collect(wo_numbers, i)
+            # Always collect coefficients
+            rebuilt = collect_const(expr.func(*args))
+            try:
+                rebuilt = Mul(*rebuilt.args)
+            except AttributeError:
+                pass
 
-            rebuilt = expr.func(w_numbers, wo_numbers)
-            return rebuilt, []
-        elif expr.is_Mul:
-            rebuilt, candidates = zip(*[run(arg) for arg in expr.args])
-            rebuilt = collect_const(expr.func(*rebuilt))
-            return rebuilt, flatten(candidates)
+            return rebuilt, ReducerMap.fromdicts(*candidates)
         elif expr.is_Equality:
-            rebuilt, candidates = zip(*[run(expr.lhs), run(expr.rhs)])
-            return expr.func(*rebuilt, evaluate=False), flatten(candidates)
+            args, candidates = zip(*[run(expr.lhs), run(expr.rhs)])
+            return expr.func(*args, evaluate=False), ReducerMap.fromdicts(*candidates)
```

Comment:
stab in the dark as I'm not quite sure what the `evaluate=False` achieves in this context... but do Increments need special handling here, too?